### DISCUSSION TITLE
Abort capture if camera hung

### DISCUSSION THREAD

**Discussion by jvinraviv** [aaronwmorris/indi-allsky#1013] [Ideas]
If the exposure is taking more than 5 minutes add the option to abort the image capture and reset the camera.
This might help if the camera itself is having issues.

This is what I have in mind (have not tested it yet and it really should be an option - not hardcoded)
`# check exposure state every 5 minutes
                if check_exposure_state < loop_start_time:
                    check_exposure_state = time.time() + 300  # next check in 5 minutes

                    camera_last_ready_s = int(loop_start_time - camera_ready_time)
                    if camera_last_ready_s > 300:
                        self._miscDb.addNotification(
                            NotificationCategory.CAMERA,
                            'last_ready',
                            'Camera last ready {0:d}s ago.  Camera might be hung.'.format(camera_last_ready_s),
                            expire=timedelta(minutes=60),
                            self.indiclient.abortCcdExposure(),
                        )`

**Comment by aaronwmorris** [comment]
I see you found the abort function I added.  :-)

I have not tested it yet, but my gut feeling is that the abort function will likely not recover from a situation where the the exposure control goes into an `Alert` status.  I believe these are usually problems inside the indiserver or with the camera driver server.  In a lot of cases, it is not sufficient to just restart indi-allsky, but you have to restart the indiserver, as well.

It certainly needs to be tested.  The real challenge is that it is also difficult to create test cases.

**Comment by jvinraviv** [comment]
At least in my case restarting indi-server seems to solve the issue...

I've wrangled it a bit and this is what I came up with - I'll let it run for a bit and let you know how it works.  Not a huge Python nerd. trying to convert from C++ to this :)

`                # check exposure state every 5 minutes
                if check_exposure_state < loop_start_time:
                    check_exposure_state = time.time() + self.exposure_max_v.value + 5  # next check in 5 minutes

                    camera_last_ready_s = int(loop_start_time - camera_ready_time)
                    if camera_last_ready_s > self.exposure_max_v.value + 5:
                        self._miscDb.addNotification(
                            NotificationCategory.CAMERA,
                            'last_ready',
                            'Camera last ready {0:d}s ago.  Camera might be hung.'.format(camera_last_ready_s),
                            expire=timedelta(minutes=60),
                        )
                        logger.error('%0.4fs EXPOSURE NOT RECEIVED IN %0.4fs.  POSSIBLE CAMERA MIGHT BE HUNG - ABORTING EXPOSURE.', self.exposure_v.value, frame_elapsed)
                        self.indiclient.abortCcdExposure()`

**Comment by jvinraviv** [comment]
And it already worked!

> Nov  6 13:28:12 allsky [ERROR] Capture-1-811/MainThread capture.saferun() #297: 0.0003s EXPOSURE NOT RECEIVED IN 1.2869s.  POSSIBLE CAMERA MIGHT BE HUNG - ABORTING EXPOSURE.

And I can see that it restarted the capture. Awesome!!!
It also didn't wait the full 5 minutes and aborted 5 seconds after the max time.

**Reply by aaronwmorris** [reply]
Very interesting!  I appreciate he feedback on this.  I will look to implement the abort method soon.  I need to review the abort method for the libcamera functionality too.

**Comment by aaronwmorris** [comment]
I have merged #1014 to abort exposures run longer than 5 minutes.  I thought about shortening it, however some cameras do not really "obey" the provided exposure value.  This is mostly a problem with libcamera where enabling auto white balance results in the camera taking 3-5 exposures during a single exposure to get proper color balance.

**Comment by jvinraviv** [comment]
Thanks!
I reached the same conclusion after seeing that the camera takes 90 seconds for a 15 second exposure.
Good news is that it ran fine throughout the night with 3 aborts.

Seems to have solved the issue for me (fingers crossed)
